### PR TITLE
Test and improve handling of `Date_Time with_timezone=False` in Postgres

### LINKED ISSUE #8049: Avoid `Inexact_Type_Coercion` when downloading a database table
The problem arises from the fact that our in-memory backend only supports `Date_Time with_timezone=True`, but the Postgres backend also allows `Date_Time with_timezone=False`. When the data is read into memory, this triggers an `Inexact_Type_Coercion` because of the difference.

- [x] We should probably just allow this conversion to happen 'silently', since it is quite common - e.g. similar situation when parsing dates in a CSV file.
- [x] We need to ensure that the timestamp is not wrongly shifted - i.e. if adding the default timezone will not mistakenly reinterpret the date.
   - Add a test where we upload timestamps in various timezones and then download them without timezone and see how it is handled.

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

- Fixes #8049
- Adds tests for handling of Date_Time upload/download in Postgres.
- Adds tests for edge cases of handling of Decimal and Binary types in Postgres.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Column_Fetcher.enso [hunk 1]
```diff
@@ -160,7 +171,9 @@ default_fetcher_for_value_type value_type =
         Value_Type.Boolean -> boolean_fetcher
         Value_Type.Time -> time_fetcher
         # We currently don't distinguish timestamps without a timezone on the Enso value side.
-        Value_Type.Date_Time _ -> date_time_fetcher
+        Value_Type.Date_Time has_timezone -> case has_timezone of
+            True  -> date_time_fetcher
+            False -> local_date_time_fetcher
```
Comment (jdunkerley) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Column_Fetcher.enso]:
```suggestion
        Value_Type.Date_Time has_timezone ->
            if has_timezone then date_time_fetcher else local_date_time_fetcher
```

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso [hunk 2]
```diff
@@ -2157,11 +2157,18 @@ type Table
...
+                expected_types.zip actual_types expected_type-> actual_type->
+                    if expected_type == actual_type then Nothing else
+                        expected_type_kind = Meta.meta expected_type . constructor
+                        actual_type_kind = Meta.meta actual_type . constructor
+                        ## We ignore simple approximations that our in-memory backend does - things like adding default
+                           timezone (because we do not have Date_Time without timezone in-memory),
+                           or changing Float32 to Float64 are silently ignored.
+                           However, bigger changes, like a Binary type column getting coerced to Mixed - _will_ still be reported.
+                        if expected_type_kind == actual_type_kind then Nothing else
+                            warnings_builder.append (Inexact_Type_Coercion.Warning expected_type actual_type)
```
Comment (radeusgd) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:2171]: Approximations like `Date_Time with_timezone=True ==> Date_Time with_timezone=False` are done without warnings.

Approximations that change the _type constructor_, eg. `Value_Type.Binary ==> Value_Type.Mixed` or `Value_Type.Decimal ==> Value_Type.Float` are still warned about.

File: distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Statement_Setter.enso [hunk 3]
```diff
@@ -40,7 +43,16 @@ fill_hole_default stmt i value = case value of
...
+    _ : Date_Time ->
+        has_timezone = case type_hint of
+            Value_Type.Date_Time with_timezone -> with_timezone
+            # We include the timezone by default
+            _ -> True
+        case has_timezone of
+            True ->
+                JDBCUtils.setZonedDateTime stmt i value
+            False ->
+                JDBCUtils.setLocalDateTime stmt i value
```
Comment (radeusgd) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Statement_Setter.enso:55]: If the expected target type is `Date_Time with_timezone=False`, we use the special `setLocalDateTime` logic - to ensure that no timezone conversions are performed - but just the local date-time part is uploaded (the timezone is simply stripped from the value).

This is the best semantics we can do for values without timezone.

File: test/Table_Tests/src/Database/Postgres_Spec.enso [hunk 4]
```diff
@@ -338,6 +341,96 @@ postgres_specific_spec connection db_name setup =
...
+                   second operand.
+                case is_in_system_tz of
+                    True ->
+                        my_dt.at_zone Time_Zone.system . time_of_day . to_display_text . should_equal "15:30:00"
+                        t2.row_count . should_equal 4
+                        t2.at "A" . to_vector . should_equal [dt1, dt1, dt1, dt1]
+                    False ->
+                        my_dt.at_zone Time_Zone.system . time_of_day . to_display_text . should_not_equal "15:30:00"
+                        t2.row_count . should_equal 0
+                        t2.at "A" . to_vector . should_equal []
```
Comment (radeusgd) [inline test/Table_Tests/src/Database/Postgres_Spec.enso:437]: This is far from ideal, because in _interpolations_ (i.e. operations like `Filter_Condition.Equal`) the `Date_Time` is by default treated as `timestamptz`. But that means it is converted using different logic than when uploading data into the Database.

1. When uploading, we have special logic that ensures that the `Date_Time` has timezone stripped, but the local-date-time it represents is preserved as-is. This is something **we want**, because it seems to be the only intuitive invariant we can keep for converting Date_Time with timezone to one without timezone.
2. When interpolating, we currently cannot have this special logic - because our operations don't know enough information about the type the `?` 'hole' should have. So it just uses whatever _value_ is in there - but the Enso `Date_Time` value _does have a timezone_. So it is sent as `timestamptz` and the standard Postgres conversion happens - which first converts it to UTC and _then_ gets the local-date-time from it - possibly shifting the local-date-time and leading to surprising behaviour.

Comment (radeusgd) [reply test/Table_Tests/src/Database/Postgres_Spec.enso:437]: Possible solutions:
1. Revert the upload logic. We could rely on the default DB logic of converting to UTC and _then_ stripping the TZ. **I don't think we should do that**, as we discussed with @jdunkerley this is rather un-intuitive.
2. Make our operations more 'smart' about the expected types. We could 'infer' that if one side of `==` has a database value of `timestamp without timezone` then the other side that is a `?` interpolation, should also be treated as such - detecting these scenarios and adding the special conversion when preparing the query. This is rather complex to implement and will likely only benefit this specific edge case. Still, we may not be able to properly infer the type in some cases.
3. We can improve our toolset allowing to more easily work with these values, without changing the default semantics.
  i. We probably should add a warning when comparing a `timestamp without timezone` with `timestamp with timezone` column or Enso value.
  ii. We should have an operation that would allow to add a timezone to a `timestamp without timezone` column, i.e. `no_tz_col.date_ops.at_zone Time_Zone.utc` which would return a `timestamptz` column with the local-date-time interpreted to be UTC. Then we can ensure that both columns are converted to `timestamptz` and the operations are well-defined.


IMHO (3) would be our best choice.

Comment (radeusgd) [reply test/Table_Tests/src/Database/Postgres_Spec.enso:437]: I added #8122 as a ticket to implement (3.i.) because I think we really need a warning in this case - otherwise users working with `timestamp without timezone` columns will be getting weird behaviour without explanation.

We can add another ticket for (3.ii.) if we decide we want it.

File: test/Table_Tests/src/Database/Postgres_Spec.enso [hunk 5]
```diff
@@ -338,6 +341,96 @@ postgres_specific_spec connection db_name setup =
             w4.requested_type . should_equal (Value_Type.Decimal precision=Nothing scale=Nothing)
             w4.actual_type . should_equal Value_Type.Float
 
+        Test.specify "should round-trip timestamptz column, preserving intsant but converting to UTC" <|
```
Comment (GregoryTravis) [inline test/Table_Tests/src/Database/Postgres_Spec.enso]:
```suggestion
        Test.specify "should round-trip timestamptz column, preserving instant but converting to UTC" <|
```

File: test/Table_Tests/src/Database/Types/Postgres_Type_Mapping_Spec.enso [hunk 6]
```diff
@@ -128,6 +129,51 @@ spec connection =
...
+            materialized_table.at "B" . value_type . should_equal Value_Type.Mixed
+            w = Problems.expect_only_warning Inexact_Type_Coercion materialized_table
+            w.requested_type . should_be_a (Value_Type.Binary ...)
+            w.actual_type . should_equal Value_Type.Mixed
+
+            # The materialized data is represented as array of signed bytes
+            ffff = [-1, -1]
+            caffee = [-54, -1, -18]
+            beef = [-66, -17]
+            materialized_table.at "B" . to_vector . should_equal [ffff, caffee, beef]
```
Comment (radeusgd) [inline test/Table_Tests/src/Database/Types/Postgres_Type_Mapping_Spec.enso:175]: In the future we may want to change the semantics of this to ensure that we use 'unsigned bytes' (stored as Enso Integer), so that the range is `0` - `255` instead of `-128` - `127`.

File: test/Table_Tests/src/Database/Postgres_Spec.enso [hunk 7]
```diff
@@ -338,6 +341,96 @@ postgres_specific_spec connection db_name setup =
...
+
+            dt1 = Date_Time.new 2022 05 04 15 30
+            dt2 = Date_Time.new 2022 05 04 15 30 zone=(Time_Zone.utc)
+            dt3 = Date_Time.new 2022 05 04 15 30 zone=(Time_Zone.parse "US/Hawaii")
+            dt4 = Date_Time.new 2022 05 04 15 30 zone=(Time_Zone.parse "Europe/Warsaw")
+            v = [dt1, dt2, dt3, dt4]
+
+            w = Problems.expect_only_warning Inexact_Type_Coercion <|
+                table.update_rows (Table.new [["A", v]]) update_action=Update_Action.Insert
+            w.requested_type . should_equal (Value_Type.Date_Time with_timezone=True)
```
Comment (GregoryTravis) [inline test/Table_Tests/src/Database/Postgres_Spec.enso]: Why would the requested type be `with_timezone=True` here?

File: test/Table_Tests/src/Database/Postgres_Spec.enso [hunk 8]
```diff
@@ -338,6 +341,96 @@ postgres_specific_spec connection db_name setup =
...
+            table.at "A" . to_vector . should_equal v_at_z
+            table.at "A" . to_vector . should_equal_tz_agnostic v
+
+            ## We also check how the timestamp column behaves with interpolations:
+               (See analogous test showing a less nice behaviour without timezones below.)
+            v.each my_dt-> Test.with_clue my_dt.to_text+" " <|
+                # Depending on test runner's timezone, the `my_dt` may be equal to 1 or 2 entries in `v`.
+                equals_tz_agnostic dt1 dt2 =
+                    dt1.at_zone tz_zero == dt2.at_zone tz_zero
+                local_equals = v.filter (equals_tz_agnostic my_dt)
```
Comment (GregoryTravis) [inline test/Table_Tests/src/Database/Postgres_Spec.enso]: Here it looks like it's comparing all the elements of `v` to each other, ignoring time zones. Why would they ever be unequal?

**Review by GregoryTravis — APPROVED** [review]

File: test/Table_Tests/src/Database/Postgres_Spec.enso (hunk 7 above)
Comment (radeusgd) [reply test/Table_Tests/src/Database/Postgres_Spec.enso]: Okay maybe I will split it into separate lines to make it clearer.
```
src_table = Table.new [["A", v]]
src_table.at "A" . value_type . should_equal (Value_Type.Date_Time with_timezone=True)
table.update_rows src_table update_action=Update_Action.Insert
```

I hope that makes it clearer.

The _source_ table requests to upload Date_Time with timezone, but the target's table actual type is without timezone.

File: test/Table_Tests/src/Database/Postgres_Spec.enso (hunk 8 above)
Comment (radeusgd) [reply test/Table_Tests/src/Database/Postgres_Spec.enso]: I will add a comment.

`equals_tz_agnostic` checks if the two date-time values represent the _same time instant_ regardless of timezone. It's not comparing the *local* date-time but the time instant.

E.g. `2023-10-20 16:26 UTC` and `2023-10-20 16:26 [Europe/Warsaw]` each have the same local date-time, but they represent different _time instants_ (differing by 2 hours). `2023-10-20 14:26 UTC` and `2023-10-20 16:26 [Europe/Warsaw]` would represent two same time instants and would be deemed `equals_tz_agnostic`.

I will rename to `is_same_time_instant`.

**Review by 4e6 — APPROVED** [review]

File: test/Table_Tests/src/Database/Postgres_Spec.enso (hunk 4 above)
Comment (radeusgd) [reply test/Table_Tests/src/Database/Postgres_Spec.enso:437]: Also added #8126 for the (3.ii.) case. We can discuss on next backlog review if we actually want to implement it, but it seems like the most likely solution.

**Review by Cassandra-Clark — APPROVED** [review]
